### PR TITLE
Add streaming inference support for Unified RNNT model

### PR DESCRIPTION
**Pull request by andrusenkoau**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add streaming inference support for a Unified RNNT model. The training scripts for the Unified RNNT model will be released later.

The Unified RNNT model is a standard FastConformer RNNT model trained jointly in both offline and streaming modes. During streaming training, we replace the full attention masks with chunked limited masks (including left, middle, and right context) and use [Dynamic Chunked Convolutions](https://arxiv.org/pdf/2304.09325) to adapt the model for streaming inference. All the other parameters are shared and the same as in offline training. 

**Collection**: [ASR]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

The Unified RNNT model supports standard offline decoding and streaming chunked inference with recomputation of the left context. Cache decoding will be implemented later. 

**Offline inference**:

```python
python ${NEMO_DIR}/examples/asr/speech_to_text_eval.py \
    model_path=${MODEL_NAME} \
    dataset_manifest=${DECODING_MANIFEST} \
    batch_size=${BATCH_SIZE} \
    output_filename=${OUTPUT_MANIFEST}
```

**Streaming inference**:

```python
python ${NEMO_DIR}/examples/asr/asr_chunked_inference/rnnt/speech_to_text_streaming_infer_rnnt.py \
    model_path=${ASR_MODEL} \
    dataset_manifest=${DECODING_MANIFEST} \
    output_filename=${OUTPUT_MANIFEST} \
    left_context_secs=${LEFT_CONTEXT} \
    chunk_secs=${CHUNK_SECS} \
    right_context_secs=${RIGHT_CONTEXT} \
    att_context_size_as_chunk=true \
    calculate_wer=true \
    calculate_rtfx=true \
    batch_size=${BATCH_SIZE}
```

During streaming inference, you can utilize flexible context parameters. However, we recommend using the following context parameters:

| Left, s | Chunk, s | Right, s | Latency (C+R), s |
| :---: | :---: | :---: | :---: |
| 5.6 | 1.04 | 1.04 | 2.08 |
| 5.6 | 0.56 | 0.56 | 1.12 |
| 5.6 | 0.16 | 0.40 | 0.56 |
| 5.6 | 0.08 | 0.24 | 0.32 |
| 5.6 | 0.08 | 0.16 | 0.24 |
| 5.6 | 0.08 | 0.08 | 0.16 |

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/losses/rnnt.py [hunk 1]
```diff
@@ -333,8 +333,7 @@ def resolve_rnnt_loss(loss_name: str, blank_idx: int, loss_kwargs: dict = None)
 class RNNTLoss(Loss):
     @property
     def input_types(self):
-        """Input types definitions for CTCLoss.
-        """
+        """Input types definitions for CTCLoss."""
```
Comment (artbataev) [inline nemo/collections/asr/losses/rnnt.py]: Please fix -> `RNNTLoss`

File: examples/asr/asr_chunked_inference/rnnt/speech_to_text_streaming_infer_rnnt.py [hunk 2]
```diff
@@ -298,6 +302,12 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
         right=context_encoder_frames.right * encoder_subsampling_factor * features_frame2audio_samples,
     )
 
+    # unified ASR model: use the att_context_size as chunk size (important for extra-low latency)
+    if asr_model.cfg.encoder.att_context_style == 'chunked_limited_with_rc' and cfg.att_context_size_as_chunk:
```
Comment (artbataev) [inline examples/asr/asr_chunked_inference/rnnt/speech_to_text_streaming_infer_rnnt.py:306]: Can we add this also to `nemo/collections/asr/inference`?

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 3]
```diff
@@ -126,6 +130,12 @@ class ConformerEncoder(NeuralModule, StreamingEncoder, Exportable, AccessMixin):
             `None` means `[(conv_kernel_size-1)//2`, `(conv_kernel_size-1)//2]`, and 'causal' means
             `[(conv_kernel_size-1), 0]`.
             Defaults to None.
+        conv_context_style (str): 'regular', 'dcc', or 'dcc_rc'
```
Comment (artbataev) [inline nemo/collections/asr/modules/conformer_encoder.py]: Please add a brief explanation of the difference between `dcc` and `dcc_rc`

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 4]
```diff
@@ -816,6 +857,32 @@ def _create_masks(self, att_context_size, padding_length, max_audio_length, offs
                         torch.le(diff_chunks, left_chunks_num), torch.ge(diff_chunks, 0)
                     )
                     att_mask = torch.logical_and(att_mask, chunked_limited_mask.unsqueeze(0))
+            elif self.att_context_style == "chunked_limited_with_rc" and sum(att_context_size) != -3:
+                assert (
+                    len(att_context_size) == 3
+                ), "att_context_size must have 3 elements: [left_context, chunk_size, right_context]"
+
+                left_context_frames = att_context_size[0]
```
Comment (artbataev) [inline nemo/collections/asr/modules/conformer_encoder.py:848]: Please, separate to a different function if possible
